### PR TITLE
Fixes #23933 - consider associated object in set_taxonomies

### PR DESCRIPTION
**Pull request by kgaikwad**

### CODE REVIEW

File: app/models/concerns/audit_extensions.rb [hunk 1]
```diff
@@ -98,7 +100,19 @@ def set_taxonomies
...
   end
+
+  def set_taxonomies_using_associated(key_name)
+    ids_arr = []
+    if associated.respond_to?(:"#{key_name}_id")
+      ids_arr = [associated.send("#{key_name}_id")].compact.uniq
+    elsif associated.respond_to?(:"#{key_name}_ids")
+      ids_arr = associated.send("#{key_name}_ids")
+    end
+    self.send("#{key_name}_ids=", ids_arr)
```
Comment (ik5) [inline app/models/concerns/audit_extensions.rb:116]: Is there a way that the `#{key_name}_id`, `"#{key_name}_ids"` and `#{key_name}_ids=` does not exists at all?

**Review by ik5 — COMMENTED** [review]
small question regarding the implementation

File: app/models/concerns/audit_extensions.rb (hunk 1 above)
Comment (kgaikwad) [reply app/models/concerns/audit_extensions.rb:116]: In `set_taxonomies` method, we already checked for `SETTINGS[:organizations_enabled] &  SETTINGS[:locations_enabled]`. 
`#{key_name}_ids=` method called on `audit` object so this method will exist for audit record.

For associated object it is good to check for methods i.e ` #{key_name}_id, "#{key_name}_ids"`.

@tbrisker, please correct me if I am missing anything.

Comment (tbrisker) [reply app/models/concerns/audit_extensions.rb:116]: What @kgaikwad said, we don't know if the associated object has 0,1, or many taxonomies which is why the `respond_to?` check is needed, but we know that audit does so `#{key_name}_ids=` is defined.

**Review by tbrisker — COMMENTED** [review]
Code looks good, can you please add a test that makes sure that this works correctly (e.g. that a nic audit records the host's taxonomies etc)?

File: app/models/concerns/audit_extensions.rb (hunk 1 above)
Comment (ik5) [reply app/models/concerns/audit_extensions.rb:116]: @tbrisker  the thing that captured my eyes here, is that if both `if` validation fails, it still send the `self.send`, so I was, but if `respond_to` failed, then the writer should also not exists

Comment (tbrisker) [reply app/models/concerns/audit_extensions.rb:116]: note that `respond_to?` is checked on `associated`, while `self` is the audit object, not the associated one.

Comment (ik5) [reply app/models/concerns/audit_extensions.rb:116]: right, sorry

**Comment by kgaikwad** [comment]
@tbrisker, 
Updated PR with the test case which ensures taxonomies changes for audit.

**Review by tbrisker — APPROVED** [review]
Thanks @kgaikwad !

**Comment by tbrisker** [comment]
@kgaikwad - can you please open a cherry-pick pr to 1.18-stable?

**Comment by kgaikwad** [comment]
@tbrisker,  sure!

**Pull request merged**